fastapi/_compat.py

```diff
@@ -72,6 +81,75 @@
             general_plain_validator_function as with_info_plain_validator_function,  # noqa: F401
         )
 
+    class GenerateJsonSchema(_GenerateJsonSchema):
+        def __init__(
+            self, by_alias: bool = True, ref_template: str = DEFAULT_REF_TEMPLATE
+        ):
+            super().__init__(by_alias=by_alias, ref_template=ref_template)
+            self.skip_null_schema = False
+
+        def nullable_schema(
+            self, schema: core_schema.NullableSchema
+        ) -> JsonSchemaValue:
+            if self.skip_null_schema:
+                return super().generate_inner(schema["schema"])
+            return super().nullable_schema(schema)
+
+        def default_schema(
+            self, schema: core_schema.WithDefaultSchema
+        ) -> JsonSchemaValue:
+            json_schema = super().default_schema(schema)
+            if (
+                self.skip_null_schema
+                and json_schema.get("default", PydanticUndefined) is None
+            ):
+                json_schema.pop("default")
+            return json_schema
+
+        def generate_definitions(
+            self,
+            inputs: Sequence[
+                Tuple[JsonSchemaKeyT, JsonSchemaMode, core_schema.CoreSchema]
+            ],
+        ) -> Tuple[
+            Dict[Tuple[JsonSchemaKeyT, JsonSchemaMode], JsonSchemaValue],
+            Dict[DefsRef, JsonSchemaValue],
+        ]:
+            # Avoid circular import - Maybe there's a better way to check if it's a Param
+            from fastapi.params import Param
+
+            if self._used:
+                raise PydanticUserError(
+                    "This JSON schema generator has already been used to generate a JSON schema. "
+                    f"You must create a new instance of {type(self).__name__} to generate a new JSON schema.",
+                    code="json-schema-already-used",
+                )
+
+            for key, mode, schema in inputs:
+                self._mode = mode
+                self.skip_null_schema = isinstance(key, ModelField) and isinstance(
+                    key.field_info, Param
+                )
+                self.generate_inner(schema)
+
+            definitions_remapping = self._build_definitions_remapping()
+
+            json_schemas_map: Dict[Tuple[JsonSchemaKeyT, JsonSchemaMode], DefsRef] = {}
+            for key, mode, schema in inputs:
+                self._mode = mode
+                self.skip_null_schema = isinstance(key, ModelField) and isinstance(
```

Comment:
@maurerle I tried out this version, and I see this custom schema is only effective when the field is a parameter in FastAPI endpoints. Can we extend this to fields of pydantic models as well?
When I modify the condition to `isinstance(key, ModelField) or isinstance(key.field_info, Param)`, it seems working for both cases

Reply:
should be working now

Reply:
Nope, this does not work out with various test cases.
Can you give an MWE of what you think is still wrong?